File: leetcode/Tree/construct_binary_tree_from_inorder_and_preorder_traversal.cpp
// https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int preorderIndex;
    unordered_map<int, int> mapping;
private:
    TreeNode* build(const vector<int>& preorder, const vector<int>& inorder, int start, int end) {
        if (start > end) {
            return nullptr;
        }

        int rootVal = preorder[preorderIndex++];
        TreeNode* root = new TreeNode(rootVal);
        int mid = mapping[rootVal];

        root->left = build(preorder, inorder, start, mid - 1);
        root->right = build(preorder, inorder, mid + 1, end);

        return root;
    }
public:
    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        mapping.clear();
        for (int i = 0; i < inorder.size(); i++) {
            mapping[inorder[i]] = i;
        }

        preorderIndex = 0;

        return build(preorder, inorder, 0, inorder.size() - 1);
    }
};

File: leetcode/Tree/construct_binary_tree_from_inorder_and_preorder_traversal.py
# https://leetcode.com/problems/construct-binary-tree-from-preorder-and-inorder-traversal/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        # print(preorder, inorder)
        if len(preorder) == 0 or len(inorder) == 0:
            return None
        center = preorder[0]
        root = TreeNode(center)
        center_inx = inorder.index(center)

        if inorder[:center_inx]:
            left_inorder = inorder[:center_inx]
            left_preorder = preorder[1: len(inorder[:center_inx]) + 1 ]
            root.left = self.buildTree(left_preorder, left_inorder)
        if inorder[center_inx+1:]:
            right_inorder = inorder[center_inx+1:]
            right_preorder = preorder[1 + len(inorder[:center_inx]) : ]
            root.right = self.buildTree(right_preorder, right_inorder)

        # print(center_inx)

        # root.left = self.buildTree(preorder[1:], inorder[:center_inx])
        # root.right = self.buildTree(preorder[2:], inorder[center_inx+1:])

        return root
    